### PR TITLE
Firestore connect

### LINKED ISSUE #16: Add Firestore as backend
Adding Firestore as a backend needs to take a few things into consideration:

- How do we handle ids?
  - [Using slugs](https://www.npmjs.com/package/url-slug)? If so, what do we do when changing title?
  - How can we keep them unique and shareable? uuid under the hood? `username/songSlug`?
- How do we handle offline sync?
  - Shouldn't be too compicated https://firebase.google.com/docs/firestore/manage-data/enable-offline
- How do we handle access and auth?
  - Do we allow anonymous access to public songs? Probably should.

__UI__:
- [x] Auth UI, https://github.com/firebase/firebaseui-web-react

__Browser syncing__:
- [x] Generate short UUID https://www.npmjs.com/package/short-uuid
- [ ] Sync data with Redux
- [x] Initialize DB connection
  - [ ] Local
  - [x] Staging
  - [ ] Production

__Dev env__
- [ ] Firebase emulator for local dev

__Extra:__
- [ ] Offline availability

### PR DESCRIPTION
**Pull request by tskarhed**
Closes #16 

- Adds hooks for using and updating data (this is used instead of thunks)
  - useSong
  - useSongs
  - useSetlist
  - useSetlists
  - useUser
  - useAddSongToSetlist
- Adds listeners for auth changes
- Loads initial state from local Firestore cache
- Changes all setlist/songs actions to UPDATE and ADD, instead of the more granular
- Adds a RESET action which empties state arrays for setlists and songs
- Adds primary usage of shortUID for identifying songs

TODO:

- [x] Setup editing and creation of new songs/setlists
- [x] Add save button to limit update requests

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -7,12 +7,9 @@ import { createStore, applyMiddleware, compose } from 'redux';
 import { Provider } from 'react-redux';
 import thunk from 'redux-thunk';
 import rootReducer from './state/reducers';
-import { loadState, saveState } from './state/localStorage';
-import throttle from 'lodash.throttle';
```
Comment (tskarhed) [inline src/index.tsx:11]: Remove throttle as a dependency

File: src/firebase/FirebaseListeners.tsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      .collection('users')
+      .doc(userId)
+      .onSnapshot({}, (userDoc) => {});
+    // Unregister
+    return () => {
+      unsubAuth();
+      unsubUser();
+    };
+  }, [userId]);
+  return <></>;
```
Comment (tskarhed) [inline src/firebase/FirebaseListeners.tsx:33]: Can probably be just a hook

File: src/state/localStorage.ts [hunk 3]
```diff
@@ -1,15 +1,36 @@
...
-  try {
-    const serializedState = localStorage.getItem('redux-state');
-    console.log(serializedState);
-    if (serializedState === null) {
-      return undefined;
-    }
-    return JSON.parse(serializedState);
-  } catch (err) {
-    return undefined;
+export const loadState = async () => {
```
Comment (tskarhed) [inline src/state/localStorage.ts:5]: Remove this file, as local sync is handled by Firebase

File: src/firebase/FirebaseListeners.tsx (hunk 2 above)
Comment (MrTengil) [reply src/firebase/FirebaseListeners.tsx:105]: What does this do tho? :)

**Comment by tskarhed** [comment]
Removed test instead of rewriting them - pretty much the entire reducer file has been refactored, so none of them are relevant anymore.

**Pull request merged**